### PR TITLE
GIT-109936: enable configurable image GC for kubelet

### LINKED ISSUE #109936: Kubelet removes non-Kubernetes images.
### What happened?

We also hit by this issue. We're using Flatcar Linux which is quite barebone so we are using a few containers to augment it. Unfortunately kubelet removes the images when they are not in active use. For example we a have a cronjob which uses an image we pulled and specially tagged. That images isn't used by any pods at all but kubelet choose to remove it under disk pressure. It is very annoying.

It would be nice to have a setting which says that "you can touch images only if the tag is starting with this prefix" or similar.

Even if you are using another Linux distribution then you might want to install something which isn't available in the distribution itself. Using a container is a very likely way to go and you don't expect that something which never belonged to the Kubernetes cluster will be removed by kubelet. It's very unintuitive.

### What did you expect to happen?

Kubelet won't remove images which aren't belongs to it.

### How can we reproduce it (as minimally and precisely as possible)?

* Pull some images from the command line.
* Fill up the local disk to create a DiskPressure "situation".
* Wait until kubelet removes the images - by the documentation it does in every five minutes.

### Anything else we need to know?

I created this new ticket because the old one (#103255) I found gets closed by k8s-ci-robot.

We are still using Kubernetes 1.21 as because of the faulty etcd-client lib used in 1.22 and 1.23 we cannot upgrade our clusters. The fix appeared in 1.24 but never backported to 1.22 or 1.23 which left us no possible upgrade paths.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:59:11Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:53:14Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
Kubernetes cluster running on EC2 instances in AWS.
</details>

### OS version

<details>

```console
$ cat /etc/os-release
NAME="Flatcar Container Linux by Kinvolk"
ID=flatcar
ID_LIKE=coreos
VERSION=3033.2.3
VERSION_ID=3033.2.3
BUILD_ID=2022-03-02-2032
PRETTY_NAME="Flatcar Container Linux by Kinvolk 3033.2.3 (Oklo)"
ANSI_COLOR="38;5;75"
HOME_URL="https://flatcar-linux.org/"
BUG_REPORT_URL="https://issues.flatcar-linux.org"
FLATCAR_BOARD="amd64-usr"
$ uname -a
Linux ip-10-8-74-48.ec2.internal 5.10.102-flatcar #1 SMP Wed Mar 2 19:40:13 -00 2022 x86_64 AMD EPYC 7571 AuthenticAMD GNU/Linux
```

</details>

### Install tools

<details>
None.
</details>

### Container runtime (CRI) and version (if applicable)

<details>
N/A
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
N/A
</details>

### PR DESCRIPTION
**Pull request by harshanarayana**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

Currently, the way the Kubernetes performs a GC operation of the Images if the GC threshold is met is partially uncontrolled. i.e It excludes the sandbox images, and performs a `minAge` check before performing the GC of the images. However, this can lead to some serious breakdown of the cluster depending on what images have been GC'ed. 

In cases such as AirGap cluster it is important that there is a provision to be able to instruct kubelet which images to exempt from GC workflow. This is required to ensure that we can retain some cluster critical images on the cluster always.

#### Which issue(s) this PR fixes:

Fixes #109936

#### Special notes for your reviewer:
```yaml
exemptByTag:
- image1:tag1
- image2:tag2
exemptByID:
- "1"
- "2"
```
Above is the format of the GC Configuration file.  The PR in current state doesn't support wildcarding the image tags. 

TODO: 

- [ ] Version the GC Configuration spec

#### Does this PR introduce a user-facing change?
Yes

```release-note
kubelet will  provide a configurable way to indicate the exemption configuration for the Container Images. This configuration can be provided either using the tags or the Images ID using suitable structure of the GC Configuration file
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
NA for now. PoC to identify if we really need a KEP as suggested by @dims 
```

### CODE REVIEW

File: pkg/kubelet/images/image_gc_manager.go [hunk 1]
```diff
@@ -149,6 +166,8 @@ type imageRecord struct {
 
 	// Pinned status of the image
 	pinned bool
+
+	exempted bool
```
Comment (ruiwen-zhao) [inline pkg/kubelet/images/image_gc_manager.go:170]: I would prefer if we can re-use `pinned` here. Having two fields for very similar purposes seems confusing to users.

Comment (harshanarayana) [reply pkg/kubelet/images/image_gc_manager.go:170]: It is kept intentionally different as the current meaning associated with `pinned` is entirely different and this is driven as a CRI attribute driven state during the GC workflow. Exempted clearly annotates that this has been excluded from GC'ing by an external configuration list provided to the Kubelet.  Merging exempted and pinned would violate the meaning associated with the attribute. If we are to merge, we could rather rename both the fields and name call it as `ignoreGC` may be ?

**Comment by endocrimes** [comment]
/remove-kind bug
/kind feature
/remove-triage accepted
/triage needs-information
/hold

This change requires a kep. We're already part way through rolling out support for the `pinned` field (https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2040-kubelet-cri#pinned-images) that should exempt images from garbage collection.

Complicating the API before it's in common use needs to be thought through in detail, rather than in a change PR.

**Comment by fedebongio** [comment]
/remove-sig api-machinery
(again)

**Comment by harshanarayana** [comment]
> Complicating the API before it's in common use needs to be thought through in detail, rather than in a change PR.

@endocrimes This PR was opened as a way to discuss the possibility of wether this change would need a MIP or not.  I am perfectly happy to start a KEP or discuss this further as required before we do any more work on this PR.

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Pull request closed without merging**